src/components/cc-kv-explorer/cc-kv-explorer.js

```diff
@@ -0,0 +1,1078 @@
+          await this._keysRef?.value?.layoutComplete;
+        } catch (e) {}
+      },
+    );
+
+    // this is for lit-virtualizer
+    this._renderKey = this._renderKey.bind(this);
+    this._onDeleteKeyButtonClick = this._onDeleteKeyButtonClick.bind(this);
+    // this is for form submission
+    this._onKeyAddFormSubmit = this._onKeyAddFormSubmit.bind(this);
+    this._onFilterFormSubmit = this._onFilterFormSubmit.bind(this);
+  }
+
+  /**
+   * Resets the add form
+   */
+  resetAddEditorForm() {
+    this._addFormRef.value?.reset();
+  }
+
+  /**
+   * Resets the hash editor add form
+   */
+  resetHashEditor() {
+    this._hashEditor.value.resetAddForm();
+  }
+
+  /**
+   * Resets the list editor add form
+   */
+  resetListEditor() {
+    this._listEditor.value.resetAddForm();
+  }
+
+  /**
+   * Resets the set editor add form
+   */
+  resetSetEditor() {
+    this._setEditor.value.resetAddForm();
+  }
+
+  /**
+   * Focuses the add key button
+   */
+  focusAddKeyButton() {
+    this._addButtonRef.value.focus();
+  }
+
+  async _onAddButtonClick() {
+    // display editor with add mode
+    this.detailState = { type: 'add', formState: { type: 'idle' } };
+
+    // unselect any selected key
+    const selectedKey = this._getSelectedKey();
+    if (selectedKey != null && this.state.type === 'loaded') {
+      this.state = {
+        type: 'loaded',
+        keys: this.state.keys.map((keyState) => {
+          if (keyState.type === 'selected') {
+            return {
+              type: 'idle',
+              key: keyState.key,
+            };
+          }
+          return keyState;
+        }),
+        total: this.state.total,
+      };
+    }
+
+    // move focus to the keyName input
+    await this.updateComplete;
+    this._addFormRef.value.keyName.focus();
+  }
+
+  /**
+   * @param {VisibilityChangedEvent} e
+   */
+  _onVisibilityChanged(e) {
+    if (this.state.type === 'loaded') {
+      if (e.last >= this.state.keys.length - 5) {
+        dispatchCustomEvent(this, 'load-more-keys');
+      }
+    }
+  }
+
+  /**
+   * @param {HTMLInputElementEvent} e
+   */
+  _onSelectedKeyChange(e) {
+    if (this.state.type !== 'loaded') {
+      return;
+    }
+
+    const keyName = e.target.value;
+    const keyState = this._findKeyState(keyName);
+
+    if (this.supportedTypes.includes(keyState.key.type)) {
+      dispatchCustomEvent(this, 'selected-key-change', keyState.key);
+    } else {
+      this.detailState = { type: 'unsupported', key: keyState.key };
+    }
+  }
+
+  async _onTypeFilterChange() {
+    // we need to wait for next update because the `ElementInternals` form data synchronization is done asynchronously (in `updated` stage)
+    await this.updateComplete;
+    this._filterFormRef.value.requestSubmit();
+  }
+
+  /**
+   * @param {FormDataMap} formData
+   */
+  _onFilterFormSubmit(formData) {
+    const filter = /** @type {CcKvKeyFilter}*/ {
+      type: formData.keyType,
+      match: formData.match,
+    };
```

Comment:
I think that I wanted to cast but I forgot the parenthesis around the `{...}`. You're right this is not needed because the `dispatchCustomEvent` accepts any type for the detail. param. 
I'll restore the right casting since it will be needed when we will use string typed events.